torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -906,6 +922,187 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
         return result
 
 
+@register_micro_gemm(
+    *generate_gemm_config(
+        VecAVX512,
+        [
+            (4, 32, 64),
+            (8, 32, 64),
+        ],
+        input_dtype=torch.bfloat16,
+        input2_dtype=torch.int8,
+        output_dtype=torch.float,
+        compute_dtype=torch.float,
+        extra_check=check_int8_woq_small_m_dim,
+    ),
+)
+class CppMicroGemmWoQSmallMDim(CppMicroGemm):
+    """
+    This class generates a macro-kernel & micro-kernel (in BLIS lexicon) for small M dimension for
+    input_dtype bfloat16 and input2_dtype int8. M is less than 16.
+    K & N must be equal to block_k and block_n, respectively.
+    These values were found to deliver good performance during tuning, and are subject to change.
+    """
+
+    TEMPLATE_ENTRY = r"""
+{{declare_kernel}} {
+    // Counter-intuitively, converting A to FP32 in the macro-kernel (instead of the micro-kernel)
+    // did not degrade performance.
+    // TODO #1: check if we can reuse the existing GEMM micro-kernel, but with prefetching.
+    // If not, check if converting elements of A inside the micro-kernel works as well.
+    // BF16 elements are converted to FP32 inside the micro-kernel one at a time.
+    // However, that approach uses less memory.
+    // TODO #2: Consider moving the following check to the new template
+    {{kernel.assert_function}}(N == {{block_n}}, "N dimension must be equal to {{block_n}}");
+    {{kernel.assert_function}}(K == {{block_k}}, "K dimension must be equal to {{block_k}}");
+    auto cvt_bf16_to_fp32 = [&](float* A_fp32, int num_rows, int m) {
+        for (int A_row = 0; A_row < num_rows; A_row++) {
+            // Convert 16 BF16 elements at a time
+            for (int vec_num = 0; vec_num < 4; vec_num++) {
+                __m512* tmp =
+                    reinterpret_cast<__m512*>(
+                        A_fp32 + 16 * vec_num + {{block_k}} * A_row
```

Comment:
Is it possible that `16 * vec_num` is larger than `block_k`? Feels like it might be wrong in this case.

Reply:
`{{block_k}}` would always be 64 & `{{block_n}}` would always be 32. These numbers were hard-coded in Intel Extension for PyTorch, and were chosen after tuning.

Reply:
Same as the `vec_num < 4` which hold the implicit assumptions that `block_k == 64`. Let's assert it to prevent implicit correctness issue.